Firestore/core/src/firebase/firestore/immutable/array_sorted_map.h

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_IMMUTABLE_ARRAY_SORTED_MAP_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_IMMUTABLE_ARRAY_SORTED_MAP_H_
+
+#include <assert.h>
+
+#include <algorithm>
+#include <array>
+#include <functional>
+#include <memory>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/immutable/map_entry.h"
+
+namespace firebase {
+namespace firestore {
+namespace immutable {
+
+/**
+ * A base class for implementing ArraySortedMap, which contains constants and
+ * types that don't depend upon the type of ArraySortedMap's type parameters.
+ */
+class ArraySortedMapBase {
+ public:
+  /**
+   * The type of size(). Note that this is not size_t specifically to save
+   * space in the TreeSortedMap implementation.
+   */
+  typedef uint32_t size_type;
+
+  /**
+   * The maximum size of an ArraySortedMap.
+   *
+   * This is the size threshold where we use a tree backed sorted map instead
+   * of an array backed sorted map. This is a more or less arbitrary chosen
+   * value, that was chosen to be large enough to fit most of object kind of
+   * Firebase data, but small enough to not notice degradation in performance
+   * for inserting and lookups. Feel free to empirically determine this
+   * constant, but don't expect much gain in real world performance.
+   */
+  static constexpr size_type kFixedSize = 25;
+};
+
+/**
+ * A thin wrapper around std::array that also maintains a dynamic size.
```

Comment:
When I first read this comment, I assumed that the class switches to dynamic memory management if std::array size is exceeded (the word 'dynamic' gave that impression). Consider elaborating a little bit to make it clear that it uses the std::array as sort of a buffer and keeps track of how much of the buffer is currently used, and, importantly, what happens if capacity is exceeded.